cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/SourceControlMetadataStore.java

```diff
@@ -0,0 +1,228 @@
+  }
+
+  /**
+   * Deletes the source control metadata associated with the specified application ID from the
+   * namespace.
+   *
+   * @param appId {@link ApplicationId} whose source control metadata is to be deleted.
+   * @throws IOException if it failed to read or delete the metadata
+   */
+  public void deleteNamespaceSourceControlMetadata(ApplicationId appId) throws IOException {
+    getNamespaceSourceControlMetadataTable().delete(getPrimaryKey(appId));
+  }
+
+  /**
+   * Deletes all rows of source control metadata within the specified namespace.
+   *
+   * @param namespace The namespace for which all source control metadata rows are to be deleted.
+   * @throws IOException if it failed to read or delete the metadata.
+   */
+  public void deleteAllRowsInNamespaceSourceControlMetadata(String namespace) throws IOException {
+    getNamespaceSourceControlMetadataTable().deleteAll(getNamespaceRange(namespace));
+  }
+
+
+  private Collection<Field<?>> getNamespaceSourceControlMetaFields(ApplicationId appId,
+      SourceControlMeta scmMeta) {
+    List<Field<?>> fields = getPrimaryKey(appId);
+    fields.add(Fields.stringField(
+        StoreDefinition.NamespaceSourceControlMetadataStore.SPECIFICATION_HASH_FIELD,
+        scmMeta == null ? "" : scmMeta.getFileHash()));
+    fields.add(
+        Fields.stringField(StoreDefinition.NamespaceSourceControlMetadataStore.COMMIT_ID_FIELD,
+            scmMeta == null ? "" : scmMeta.getCommitId()));
+    fields.add(
+        Fields.longField(StoreDefinition.NamespaceSourceControlMetadataStore.LAST_MODIFIED_FIELD,
+            scmMeta == null ? 0L : scmMeta.getLastSyncedAt().toEpochMilli()));
+    fields.add(
+        Fields.booleanField(StoreDefinition.NamespaceSourceControlMetadataStore.IS_SYNCED_FIELD,
+            scmMeta == null ? false : true));
+    return fields;
+  }
+
+  private List<Field<?>> getPrimaryKey(ApplicationId appId) {
+    List<Field<?>> primaryKey = new ArrayList<>();
+    primaryKey.add(
+        Fields.stringField(StoreDefinition.NamespaceSourceControlMetadataStore.NAMESPACE_FIELD,
+            appId.getNamespace()));
+    primaryKey.add(
+        Fields.stringField(StoreDefinition.NamespaceSourceControlMetadataStore.TYPE_FIELD,
+            appId.getEntityType().toString()));
+    primaryKey.add(
+        Fields.stringField(StoreDefinition.NamespaceSourceControlMetadataStore.NAME_FIELD,
+            appId.getEntityName()));
+    return primaryKey;
+  }
+
+  private Range getNamespaceRange(String namespaceId) {
+    return Range.singleton(
+        ImmutableList.of(
+            Fields.stringField(StoreDefinition.AppMetadataStore.NAMESPACE_FIELD, namespaceId)));
+  }
+
+  /**
+   * Deletes all rows from the namespace source control metadata table. Only to be used in testing.
+   *
+   * @throws IOException If an I/O error occurs while deleting the metadata.
+   */
+  @VisibleForTesting
+  public void deleteNamespaceSourceControlMetadataTable() throws IOException {
```

Comment:
usually things that are only visible for testing can be package private, as tests are normally only within the package.

Reply:
Noted. Fixed

Reply:
Noted. Fixed